express_relay/programs/express_relay/src/lib.rs

```diff
@@ -0,0 +1,316 @@
+        let rent_relayer_fee_receiver = Rent::default().minimum_balance(0).max(1);
+        let amount_relayer: u64;
+        if balance_relayer_fee_receiver >= rent_relayer_fee_receiver {
+            amount_relayer = fee_relayer;
+        } else {
+            amount_relayer = fee_relayer + (rent_relayer_fee_receiver - balance_relayer_fee_receiver);
+        }
+
+        transfer_lamports_cpi(
+            &searcher.to_account_info(),
+            &protocol_fee_receiver.to_account_info(),
+            amount_protocol,
+            ctx.accounts.system_program.to_account_info()
+        )?;
+        transfer_lamports_cpi(
+            &searcher.to_account_info(),
+            &relayer_fee_receiver.to_account_info(),
+            amount_relayer,
+            ctx.accounts.system_program.to_account_info()
+        )?;
+        // send the remaining balance from the bid to the express relay metadata account
+        transfer_lamports_cpi(
+            &searcher.to_account_info(),
+            &express_relay_metadata.to_account_info(),
+            bid_amount.saturating_sub(fee_protocol).saturating_sub(fee_relayer),
+            ctx.accounts.system_program.to_account_info()
+        )?;
+
+        Ok(())
+    }
+
+    pub fn check_permission(ctx: Context<CheckPermission>) -> Result<()> {
+        let num_instructions = read_u16(&mut 0, &ctx.accounts.sysvar_instructions.data.borrow()).map_err(|_| ProgramError::InvalidInstructionData)?;
+        for index in 0..num_instructions {
+            let ix = load_instruction_at_checked(index.into(), &ctx.accounts.sysvar_instructions)?;
+
+            if ix.program_id != crate::id() {
+                continue;
+            }
+            let expected_discriminator = sighash("global", "permission");
+            if ix.data[0..8] != expected_discriminator {
+                continue;
+            }
+
+            if ix.accounts[2].pubkey != *ctx.accounts.permission.key {
+                continue;
+            }
+
+            if ix.accounts[3].pubkey == *ctx.accounts.protocol.key {
+                return Ok(());
+            }
+        }
+
+        return err!(ExpressRelayError::InvalidPermissioning);
+    }
+
+    pub fn withdraw_fees(ctx: Context<WithdrawFees>) -> Result<()> {
+        let admin = &ctx.accounts.admin;
+        let express_relay_metadata = &ctx.accounts.express_relay_metadata;
+
+        let express_relay_metadata_account_info = express_relay_metadata.to_account_info();
+        let rent_express_relay_metadata = Rent::default().minimum_balance(express_relay_metadata_account_info.data_len()).max(1);
+
+        if express_relay_metadata_account_info.lamports() <= rent_express_relay_metadata {
+            return err!(ExpressRelayError::InsufficientWithdrawalFunds);
+        }
+
+        let amount = express_relay_metadata.to_account_info().lamports() - rent_express_relay_metadata;
+        transfer_lamports(&express_relay_metadata_account_info, &admin.to_account_info(), amount)
```

Comment:
I'm not sure how convenient it is to transfer the funds to the admin account. In case of governance, this is probably a PDA and it might not be easy to transfer it from there. The easier way is to just accept another account here and use that instead